File: frontend/src/lib/reports.ts
/**
 * Contrato de los informes (fase 15 — ARCHITECTURE.md §13).
 *
 * Espeja /api/v1/reports: consultas de SOLO LECTURA bajo ``reports.view``,
 * con rango temporal OBLIGATORIO (``from``/``to``, techo de 366 días en el
 * servidor), agregaciones en SQL y páginas de hasta 200 filas — el mostrador
 * nunca descarga la tabla entera. El dinero viaja como string con 2 decimales
 * (§3); los importes de devoluciones llegan con signo (negativo en el
 * resumen, positivo en forma de pago, igual que el informe Z de caja).
 *
 * Como en schemas.ts, cualquier deriva del contrato falla AQUÍ (Zod) y no
 * pinta cifras corruptas.
 */

import { z } from 'zod';
import { apiFetch } from './api';

const uuid = z.string().uuid();
const money = z.string().regex(/^-?\d{1,10}\.\d{2}$/, 'importe sin formato de contrato');
const qty = z.string().regex(/^-?\d{1,6}(\.\d{1,3})?$/, 'cantidad sin formato de contrato');
const isoDate = z.string().regex(/^\d{4}-\d{2}-\d{2}$/);
// Los timestamps llegan ISO-8601; el formato exacto lo fija el servidor.
const isoDateTime = z.string().min(10);

export const REPORTS_INTERVALS = ['hour', 'day', 'week', 'month'] as const;
export type ReportInterval = (typeof REPORTS_INTERVALS)[number];

export const REPORTS_BASE = '/api/v1/reports';

/** Rango temporal obligatorio de toda consulta de informes (§13). */
export interface ReportRange {
  from: Date;
  to: Date;
  terminalId?: string;
}

/** Página pedida al servidor (techo del backend: limit ≤ 200). */
export interface ReportPage {
  limit: number;
  offset: number;
}

export interface Paged<T> {
  items: T[];
  total: number;
  limit: number;
  offset: number;
}

// ---------------------------------------------------------------------------
// Esquemas de respuesta
// ---------------------------------------------------------------------------
export const ticketRowSchema = z.object({
  id: uuid,
  order_id: uuid,
  terminal_id: uuid,
  series: z.string(),
  number: z.number().int(),
  doc_number: z.string().nullable(),
  printed_at: isoDateTime.nullable(),
  reprint_count: z.number().int(),
  created_at: isoDateTime,
  total_amount: money,
  user_id: uuid.nullable(),
});
export const ticketPageSchema = paged(ticketRowSchema);

export const invoiceRowSchema = z.object({
  id: uuid,
  customer_id: uuid,
  series: z.string(),
  year: z.number().int(),
  number: z.number().int(),
  status: z.string(),
  issue_date: isoDate,
  total_base: money,
  total_tax: money,
  total_amount: money,
  rectified_invoice_id: uuid.nullable(),
  voided_at: isoDateTime.nullable(),
  void_reason: z.string().nullable(),
  created_at: isoDateTime,
  doc_number: z.string().nullable(),
});
export const invoicePageSchema = paged(invoiceRowSchema);

/** Factura detallada: mismo render que /documents, aquí bajo reports.view. */
export const invoiceDetailSchema = z.object({
  id: uuid,
  customer_id: uuid,
  series: z.string(),
  year: z.number().int(),
  number: z.number().int(),
  doc_number: z.string(),
  status: z.string(),
  issue_date: isoDate,
  total_base: money,
  total_tax: money,
  total_amount: money,
  rectified_invoice_id: uuid.nullable(),
  voided_at: isoDateTime.nullable(),
  void_reason: z.string().nullable(),
  created_at: isoDateTime,
  payload: z.record(z.unknown()),
  lines: z.array(z.object({ order_id: uuid })),
});

export const cashSessionSchema = z.object({
  id: uuid,
  terminal_id: uuid,
  opened_by: uuid.nullable(),
  closed_by: uuid.nullable(),
  opening_amount: money,
  expected_amount: money.nullable(),
  counted_amount: money.nullable(),
  difference: money.nullable(),
  status: z.enum(['open', 'closed']),
  opened_at: isoDateTime,
  closed_at: isoDateTime.nullable(),
});
export const closurePageSchema = paged(cashSessionSchema);

export const taxLineSchema = z.object({ tax_rate: money, base: money, total: money });

export const summarySchema = z.object({
  sales_count: z.number().int(),
  sales_amount: money,
  refunds_count: z.number().int(),
  refunds_amount: money, // negativo
  voided_count: z.number().int(),
  net_amount: money,
  average_ticket: money,
  tax_breakdown: z.array(taxLineSchema),
});

export const productRowSchema = z.object({
  product_id: uuid.nullable(),
  name: z.string(),
  orders: z.number().int(),
  quantity: qty.nullable(),
  base: money,
  total: money,
});
export const productPageSchema = paged(productRowSchema);

export const categoryRowSchema = z.object({
  category_id: uuid.nullable(),
  name: z.string(),
  orders: z.number().int(),
  quantity: qty.nullable(),
  base: money,
  total: money,
});
export const categoryPageSchema = paged(categoryRowSchema);

export const waiterRowSchema = z.object({
  user_id: uuid,
  username: z.string(),
  full_name: z.string().nullable(),
  sales_count: z.number().int(),
  refunds_count: z.number().int(),
  total: money,
});
export const waiterPageSchema = paged(waiterRowSchema);

export const paymentMethodRowSchema = z.object({
  code: z.string(),
  kind: z.string(),
  sales_count: z.number().int(),
  sales_amount: money,
  refunds_count: z.number().int(),
  refunds_amount: money, // positivo, convención del informe Z
});
export const paymentMethodPageSchema = paged(paymentMethodRowSchema);

export const periodRowSchema = z.object({
  bucket: isoDateTime,
  sales_count: z.number().int(),
  sales_amount: money,
  refunds_count: z.number().int(),
  refunds_amount: money,
});
export const periodPageSchema = paged(periodRowSchema);

function paged<T extends z.ZodTypeAny>(item: T) {
  return z.object({
    items: z.array(item),
    total: z.number().int(),
    limit: z.number().int(),
    offset: z.number().int(),
  });
}

export type TicketReportRow = z.infer<typeof ticketRowSchema>;
export type TicketReportPage = z.infer<typeof ticketPageSchema>;
export type InvoiceReportRow = z.infer<typeof invoiceRowSchema>;
export type InvoiceReportPage = z.infer<typeof invoicePageSchema>;
export type InvoiceDetail = z.infer<typeof invoiceDetailSchema>;
export type CashClosure = z.infer<typeof cashSessionSchema>;
export type ClosurePage = z.infer<typeof closurePageSchema>;
export type SummaryReport = z.infer<typeof summarySchema>;
export type ProductReportRow = z.infer<typeof productRowSchema>;
export type ProductReportPage = z.infer<typeof productPageSchema>;
export type CategoryReportRow = z.infer<typeof categoryRowSchema>;
export type CategoryReportPage = z.infer<typeof categoryPageSchema>;
export type WaiterReportRow = z.infer<typeof waiterRowSchema>;
export type WaiterReportPage = z.infer<typeof waiterPageSchema>;
export type PaymentMethodReportRow = z.infer<typeof paymentMethodRowSchema>;
export type PaymentMethodReportPage = z.infer<typeof paymentMethodPageSchema>;
export type PeriodReportRow = z.infer<typeof periodRowSchema>;
export type PeriodReportPage = z.infer<typeof periodPageSchema>;

// ---------------------------------------------------------------------------
// Query string y fetchers
// ---------------------------------------------------------------------------
/** ``from``/``to`` son obligatorios: el backend rechaza consultas sin rango. */
export function rangeParams(range: ReportRange, page?: ReportPage): URLSearchParams {
  const params = new URLSearchParams();
  params.set('from', range.from.toISOString());
  params.set('to', range.to.toISOString());
  if (range.terminalId) params.set('terminal_id', range.terminalId);
  if (page) {
    params.set('limit', String(page.limit));
    params.set('offset', String(page.offset));
  }
  return params;
}

async function getPage<T>(
  path: string,
  params: URLSearchParams,
  schema: z.ZodType<T>,
): Promise<T> {
  const raw = await apiFetch<unknown>(`${path}?${params.toString()}`);
  return schema.parse(raw);
}

export function fetchTickets(range: ReportRange, page: ReportPage) {
  return getPage(`${REPORTS_BASE}/tickets`, rangeParams(range, page), ticketPageSchema);
}

/** Facturas: filtros por fecha de emisión (día, "YYYY-MM-DD") + estado y serie. */
export function invoiceParams(
  opts: { from: string; to: string; status?: string; series?: string },
  page: ReportPage,
): URLSearchParams {
  const params = new URLSearchParams();
  params.set('from', opts.from);
  params.set('to', opts.to);
  if (opts.status) params.set('status', opts.status);
  if (opts.series) params.set('series', opts.series);
  params.set('limit', String(page.limit));
  params.set('offset', String(page.offset));
  return params;
}

export function fetchInvoices(
  opts: { from: string; to: string; status?: string; series?: string },
  page: ReportPage,
) {
  return getPage(`${REPORTS_BASE}/invoices`, invoiceParams(opts, page), invoicePageSchema);
}

export async function fetchInvoiceDetail(invoiceId: string): Promise<InvoiceDetail> {
  const raw = await apiFetch<unknown>(`${REPORTS_BASE}/invoices/${invoiceId}`);
  return invoiceDetailSchema.parse(raw);
}

export function fetchCashClosures(range: ReportRange, page: ReportPage) {
  return getPage(`${REPORTS_BASE}/cash-closures`, rangeParams(range, page), closurePageSchema);
}

export function fetchSummary(range: ReportRange) {
  return getPage(`${REPORTS_BASE}/stats/summary`, rangeParams(range), summarySchema);
}

export function fetchByProduct(range: ReportRange, page: ReportPage) {
  return getPage(`${REPORTS_BASE}/stats/by-product`, rangeParams(range, page), productPageSchema);
}

export function fetchByCategory(range: ReportRange, page: ReportPage) {
  return getPage(`${REPORTS_BASE}/stats/by-category`, rangeParams(range, page), categoryPageSchema);
}

export function fetchByWaiter(range: ReportRange, page: ReportPage) {
  return getPage(`${REPORTS_BASE}/stats/by-waiter`, rangeParams(range, page), waiterPageSchema);
}

export function fetchByPaymentMethod(range: ReportRange, page: ReportPage) {
  return getPage(
    `${REPORTS_BASE}/stats/by-payment-method`,
    rangeParams(range, page),
    paymentMethodPageSchema,
  );
}

export function fetchByPeriod(range: ReportRange, interval: ReportInterval, page: ReportPage) {
  const params = rangeParams(range, page);
  params.set('interval', interval);
  return getPage(`${REPORTS_BASE}/stats/by-period`, params, periodPageSchema);
}

File: backend/app/api/v1/reports.py
"""Informes (fase 15) — ``/api/v1/reports``, todo bajo ``reports.view``.

Capa de consulta: tickets emitidos, facturas (listado + detalle), cierres Z
pasados y estadísticas del periodo. Reglas de la casa:

- ``from``/``to`` OBLIGATORIOS en todas las consultas y acotados a 366 días
  (§13: los informes nunca escanean la tabla entera ni cargan millones de
  filas en el cliente).
- Agregaciones en SQL y páginas de hasta 200 filas (``limit``/``offset``);
  ``total`` cuenta filas o grupos, no órdenes crudas.
- Dinero como string con 2 decimales (§3); métricas derivadas (venta neta,
  ticket medio) calculadas en el backend.
- Sin auditoría ni eventos: un informe no muta estado (como los informes X/Z
  de caja).
"""

from datetime import date, datetime
from decimal import Decimal
from typing import Annotated, Literal
from uuid import UUID

from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel

from app.api.dependencies import CurrentUser, DbSession, require_permission
from app.api.v1.cash import CashSessionResponse, _session_response
from app.api.v1.documents import _invoice_response
from app.db.enums import InvoiceStatus
from app.services import reports as service

router = APIRouter(prefix="/reports", tags=["reports"])
_reports = [Depends(require_permission("reports.view"))]

DEFAULT_LIMIT = 50
MAX_LIMIT = 200

# Tramos temporales admitidos por el informe por periodo (date_trunc en SQL).
Interval = Literal["hour", "day", "week", "month"]


def _money(value: Decimal) -> str:
    return format(value, "f")


# ---------------------------------------------------------------------------
# Paginación común de informes
# ---------------------------------------------------------------------------
def _limit_offset(
    limit: Annotated[int, Query(ge=1, le=MAX_LIMIT)] = DEFAULT_LIMIT,
    offset: Annotated[int, Query(ge=0)] = 0,
) -> tuple[int, int]:
    return limit, offset


LimitOffset = Annotated[tuple[int, int], Depends(_limit_offset)]


# ---------------------------------------------------------------------------
# Respuestas
# ---------------------------------------------------------------------------
class TicketReportRow(BaseModel):
    id: UUID
    order_id: UUID
    terminal_id: UUID
    series: str
    number: int
    doc_number: str | None
    printed_at: datetime | None
    reprint_count: int
    created_at: datetime
    total_amount: str
    user_id: UUID | None


class TicketListResponse(BaseModel):
    items: list[TicketReportRow]
    total: int
    limit: int
    offset: int


class InvoiceReportRow(BaseModel):
    id: UUID
    customer_id: UUID
    series: str
    year: int
    number: int
    status: str
    issue_date: date
    total_base: str
    total_tax: str
    total_amount: str
    rectified_invoice_id: UUID | None
    voided_at: datetime | None
    void_reason: str | None
    created_at: datetime
    doc_number: str | None


class InvoiceListResponse(BaseModel):
    items: list[InvoiceReportRow]
    total: int
    limit: int
    offset: int


class ClosureListResponse(BaseModel):
    """Cierres Z pasados: la fila de sesión con el cuadre congelado."""

    items: list[CashSessionResponse]
    total: int
    limit: int
    offset: int


class TaxLineResponse(BaseModel):
    tax_rate: str
    base: str
    total: str


class SummaryResponse(BaseModel):
    sales_count: int
    sales_amount: str
    refunds_count: int
    refunds_amount: str  # negativo
    voided_count: int
    net_amount: str
    average_ticket: str
    tax_breakdown: list[TaxLineResponse]


class ProductRowResponse(BaseModel):
    product_id: UUID | None
    name: str
    orders: int
    quantity: str | None
    base: str
    total: str


class ProductListResponse(BaseModel):
    items: list[ProductRowResponse]
    total: int
    limit: int
    offset: int


class CategoryRowResponse(BaseModel):
    category_id: UUID | None
    name: str
    orders: int
    quantity: str | None
    base: str
    total: str


class CategoryListResponse(BaseModel):
    items: list[CategoryRowResponse]
    total: int
    limit: int
    offset: int


class WaiterRowResponse(BaseModel):
    user_id: UUID
    username: str
    full_name: str | None
    sales_count: int
    refunds_count: int
    total: str


class WaiterListResponse(BaseModel):
    items: list[WaiterRowResponse]
    total: int
    limit: int
    offset: int


class PaymentMethodRowResponse(BaseModel):
    code: str
    kind: str
    sales_count: int
    sales_amount: str
    refunds_count: int
    refunds_amount: str


class PaymentMethodListResponse(BaseModel):
    items: list[PaymentMethodRowResponse]
    total: int
    limit: int
    offset: int


class PeriodRowResponse(BaseModel):
    bucket: datetime
    sales_count: int
    sales_amount: str
    refunds_count: int
    refunds_amount: str


class PeriodListResponse(BaseModel):
    items: list[PeriodRowResponse]
    total: int
    limit: int
    offset: int


# ---------------------------------------------------------------------------
# Tickets emitidos
# ---------------------------------------------------------------------------
@router.get("/tickets", dependencies=_reports)
async def list_tickets(
    session: DbSession,
    _user: CurrentUser,
    since: Annotated[datetime, Query(alias="from")],
    until: Annotated[datetime, Query(alias="to")],
    terminal_id: UUID | None = None,
    limits: LimitOffset = (DEFAULT_LIMIT, 0),
) -> TicketListResponse:
    limit, offset = limits
    rows, total = await service.tickets_page(
        session,
        issued_from=since,
        issued_to=until,
        terminal_id=terminal_id,
        limit=limit,
        offset=offset,
    )
    return TicketListResponse(
        items=[
            TicketReportRow(
                id=row.id,
                order_id=row.order_id,
                terminal_id=row.terminal_id,
                series=row.series,
                number=row.number,
                doc_number=row.doc_number,
                printed_at=row.printed_at,
                reprint_count=row.reprint_count,
                created_at=row.created_at,
                total_amount=_money(row.total_amount),
                user_id=row.user_id,
            )
            for row in rows
        ],
        total=total,
        limit=limit,
        offset=offset,
    )


# ---------------------------------------------------------------------------
# Facturas
# ---------------------------------------------------------------------------
@router.get("/invoices", dependencies=_reports)
async def list_invoices(
    session: DbSession,
    _user: CurrentUser,
    since: Annotated[date, Query(alias="from")],
    until: Annotated[date, Query(alias="to")],
    status: InvoiceStatus | None = None,
    series: str | None = None,
    limits: LimitOffset = (DEFAULT_LIMIT, 0),
) -> InvoiceListResponse:
    limit, offset = limits
    rows, total = await service.invoices_page(
        session,
        issue_from=since,
        issue_to=until,
        status=status,
        series=series,
        limit=limit,
        offset=offset,
    )
    return InvoiceListResponse(
        items=[
            InvoiceReportRow(
                id=row.id,
                customer_id=row.customer_id,
                series=row.series,
                year=row.year,
                number=row.number,
                status=row.status.value,
                issue_date=row.issue_date,
                total_base=_money(row.total_base),
                total_tax=_money(row.total_tax),
                total_amount=_money(row.total_amount),
                rectified_invoice_id=row.rectified_invoice_id,
                voided_at=row.voided_at,
                void_reason=row.void_reason,
                created_at=row.created_at,
                doc_number=row.doc_number,
            )
            for row in rows
        ],
        total=total,
        limit=limit,
        offset=offset,
    )


@router.get("/invoices/{invoice_id}", dependencies=_reports)
async def invoice_detail(
    session: DbSession, _user: CurrentUser, invoice_id: UUID
) -> object:
    """Factura detallada: el snapshot congelado con el que se emitió — el
    mismo render que la API de documentos, aquí bajo ``reports.view``."""

    invoice, lines = await service.invoice_detail(session, invoice_id)
    return _invoice_response(invoice, lines)


# ---------------------------------------------------------------------------
# Cierres pasados (Z)
# ---------------------------------------------------------------------------
@router.get("/cash-closures", dependencies=_reports)
async def list_cash_closures(
    session: DbSession,
    _user: CurrentUser,
    since: Annotated[datetime, Query(alias="from")],
    until: Annotated[datetime, Query(alias="to")],
    terminal_id: UUID | None = None,
    limits: LimitOffset = (DEFAULT_LIMIT, 0),
) -> ClosureListResponse:
    limit, offset = limits
    rows, total = await service.closures_page(
        session,
        closed_from=since,
        closed_to=until,
        terminal_id=terminal_id,
        limit=limit,
        offset=offset,
    )
    return ClosureListResponse(
        items=[_session_response(row) for row in rows],
        total=total,
        limit=limit,
        offset=offset,
    )


# ---------------------------------------------------------------------------
# Estadísticas
# ---------------------------------------------------------------------------
@router.get("/stats/summary", dependencies=_reports)
async def stats_summary(
    session: DbSession,
    _user: CurrentUser,
    since: Annotated[datetime, Query(alias="from")],
    until: Annotated[datetime, Query(alias="to")],
    terminal_id: UUID | None = None,
) -> SummaryResponse:
    s = await service.summary(
        session, paid_from=since, paid_to=until, terminal_id=terminal_id
    )
    return SummaryResponse(
        sales_count=s.sales_count,
        sales_amount=_money(s.sales_amount),
        refunds_count=s.refunds_count,
        refunds_amount=_money(s.refunds_amount),
        voided_count=s.voided_count,
        net_amount=_money(s.net_amount),
        average_ticket=_money(s.average_ticket),
        tax_breakdown=[
            TaxLineResponse(
                tax_rate=_money(t.tax_rate),
                base=_money(t.base),
                total=_money(t.total),
            )
            for t in s.tax_breakdown
        ],
    )


@router.get("/stats/by-product", dependencies=_reports)
async def stats_by_product(
    session: DbSession,
    _user: CurrentUser,
    since: Annotated[datetime, Query(alias="from")],
    until: Annotated[datetime, Query(alias="to")],
    terminal_id: UUID | None = None,
    limits: LimitOffset = (DEFAULT_LIMIT, 0),
) -> ProductListResponse:
    limit, offset = limits
    rows, total = await service.by_product(
        session, paid_from=since, paid_to=until, terminal_id=terminal_id,
        limit=limit, offset=offset,
    )
    return ProductListResponse(
        items=[
            ProductRowResponse(
                product_id=row.product_id,
                name=row.name,
                orders=row.orders,
                quantity=str(row.quantity) if row.quantity is not None else None,
                base=_money(row.base),
                total=_money(row.total),
            )
            for row in rows
        ],
        total=total,
        limit=limit,
        offset=offset,
    )


@router.get("/stats/by-category", dependencies=_reports)
async def stats_by_category(
    session: DbSession,
    _user: CurrentUser,
    since: Annotated[datetime, Query(alias="from")],
    until: Annotated[datetime, Query(alias="to")],
    terminal_id: UUID | None = None,
    limits: LimitOffset = (DEFAULT_LIMIT, 0),
) -> CategoryListResponse:
    limit, offset = limits
    rows, total = await service.by_category(
        session, paid_from=since, paid_to=until, terminal_id=terminal_id,
        limit=limit, offset=offset,
    )
    return CategoryListResponse(
        items=[
            CategoryRowResponse(
                category_id=row.category_id,
                name=row.name,
                orders=row.orders,
                quantity=str(row.quantity) if row.quantity is not None else None,
                base=_money(row.base),
                total=_money(row.total),
            )
            for row in rows
        ],
        total=total,
        limit=limit,
        offset=offset,
    )


@router.get("/stats/by-waiter", dependencies=_reports)
async def stats_by_waiter(
    session: DbSession,
    _user: CurrentUser,
    since: Annotated[datetime, Query(alias="from")],
    until: Annotated[datetime, Query(alias="to")],
    terminal_id: UUID | None = None,
    limits: LimitOffset = (DEFAULT_LIMIT, 0),
) -> WaiterListResponse:
    limit, offset = limits
    rows, total = await service.by_waiter(
        session, paid_from=since, paid_to=until, terminal_id=terminal_id,
        limit=limit, offset=offset,
    )
    return WaiterListResponse(
        items=[
            WaiterRowResponse(
                user_id=row.user_id,
                username=row.username,
                full_name=row.full_name,
                sales_count=row.sales_count,
                refunds_count=row.refunds_count,
                total=_money(row.total),
            )
            for row in rows
        ],
        total=total,
        limit=limit,
        offset=offset,
    )


@router.get("/stats/by-payment-method", dependencies=_reports)
async def stats_by_payment_method(
    session: DbSession,
    _user: CurrentUser,
    since: Annotated[datetime, Query(alias="from")],
    until: Annotated[datetime, Query(alias="to")],
    terminal_id: UUID | None = None,
    limits: LimitOffset = (DEFAULT_LIMIT, 0),
) -> PaymentMethodListResponse:
    limit, offset = limits
    rows, total = await service.by_payment_method(
        session, paid_from=since, paid_to=until, terminal_id=terminal_id,
        limit=limit, offset=offset,
    )
    return PaymentMethodListResponse(
        items=[
            PaymentMethodRowResponse(
                code=row.code,
                kind=row.kind.value,
                sales_count=row.sales_count,
                sales_amount=_money(row.sales_amount),
                refunds_count=row.refunds_count,
                refunds_amount=_money(row.refunds_amount),
            )
            for row in rows
        ],
        total=total,
        limit=limit,
        offset=offset,
    )


@router.get("/stats/by-period", dependencies=_reports)
async def stats_by_period(
    session: DbSession,
    _user: CurrentUser,
    since: Annotated[datetime, Query(alias="from")],
    until: Annotated[datetime, Query(alias="to")],
    interval: Interval = "day",
    terminal_id: UUID | None = None,
    limits: LimitOffset = (DEFAULT_LIMIT, 0),
) -> PeriodListResponse:
    limit, offset = limits
    rows, total = await service.by_period(
        session, paid_from=since, paid_to=until, terminal_id=terminal_id,
        interval=interval, limit=limit, offset=offset,
    )
    return PeriodListResponse(
        items=[
            PeriodRowResponse(
                bucket=row.bucket,
                sales_count=row.sales_count,
                sales_amount=_money(row.sales_amount),
                refunds_count=row.refunds_count,
                refunds_amount=_money(row.refunds_amount),
            )
            for row in rows
        ],
        total=total,
        limit=limit,
        offset=offset,
    )
